### PR TITLE
Adds enumerables class

### PR DESCRIPTION
**Pull request by tiagoit**
## About The Project

Implementation of the most used methods on the Enumerable Ruby class.

### Built With

* [Ruby](https://www.ruby-lang.org)
* [VS Code](https://code.visualstudio.com/)

## Getting Started

Uncomment the examples and see the results on IRB console.

## License

Distributed under the MIT License. See `LICENSE` for more information.

## Contact

Project Link: [https://github.com/ferreirati/mv-ruby-02-enumerables](https://github.com/ferreirati/mv-ruby-02-enumerables)

- **Tiago Ferreira** - [@ferreirati](https://github.com/ferreirati)

### CODE REVIEW

File: src/enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    length.times do |i|
+      yield(self[i])
```
Comment (samgaco) [inline src/enumerable.rb]: ### 1. Should return an enumerator when not passing arguments or blocks

example :arrow_down_small: 
```

irb(main):061:0> [1,2,3].each
=> #<Enumerator: [1, 2, 3]:each>
```

right now we are getting:
```
irb(main):062:0> [1,2,3].my_each
....
LocalJumpError (no block given (yield))

```

File: src/enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    end
+  end
+
+  def my_each_with_index
+    length.times do |i|
+      yield(self[i], i)
```
Comment (samgaco) [inline src/enumerable.rb]: #### 2. Should return an enumerator when not passing arguments or blocks (same problem as in my_each)

File: src/enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    end
+  end
+
+  def my_select
+    selected_items = []
```
Comment (samgaco) [inline src/enumerable.rb]: #### 3. Should return an enumerator when not passing arguments or blocks (same problem as in my_each)

File: src/enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    my_each do |item|
+      selected_items << item if yield(item)
+    end
+    selected_items
+  end
+
+  def my_all?
+    my_each { |item| return false unless yield(item) }
+    true
+  end
```
Comment (samgaco) [inline src/enumerable.rb]: #### 4. when no passing any arguments, it should evaluate if all the members of the connection are true, example:

```
irb(main):065:0> [1,2,3].all?
=> true
```

#### 5. when passing a class, this method should return true if all the members in the evaluated collection ,[1,2,3] for example, are members of that class:

expected:

```
irb(main):063:0> [1,2,3].all?(Integer)
=> true
```

got:

```
irb(main):064:0> [1,2,3].my_all?(Integer)
...
ArgumentError (wrong number of arguments (given 1, expected 0))
``` 

#### 6. when passing a regular expression, this method should return true if all the members in the evaluated collection are members of that class:

```
irb(main):068:0> ["fj","ooo"].all?(/[a-z]/)
=> true
```

#### 7. same deal when passing a pattern:

```
irb(main):069:0> [1,1,1].all?(1)
=> true
irb(main):070:0> [1,2,1].all?(1)
=> false
irb(main):071:0> ["o","o"].all?("o")
=> true
irb(main):072:0> ["o","e"].all?("o")
=> false
```

File: src/enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+  def my_any?
+    my_each { |item| return true if yield(item) }
+    false
+  end
```
Comment (samgaco) [inline src/enumerable.rb]: #### 8. , 9. , 10. , 11. 
:recycle:  All the listed behaviors in points 4,5,6 and 7 for the method my_all? are failing here too.

File: src/enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+  def my_none?
+    my_each { |item| return false if yield(item) }
+    true
+  end
```
Comment (samgaco) [inline src/enumerable.rb]: #### 12. , 13. , 14. , 15. 
:recycle:  All the listed behaviors in points 4,5,6, 7,8,9,10,11  for the method my_all? and my_any? are failing here too.

File: src/enumerable.rb [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+  def my_count
+    count = 0
+    my_each { |item| count += 1 if yield(item) }
```
Comment (samgaco) [inline src/enumerable.rb]: #### 16.  Should return the number of elements in a collection if no block is given:

```
irb(main):076:0> [1,2,3].count
=> 3
```

#### 17.  If we pass an argument with a pattern, it should count the number of times that such given pattern appears in the collection, example:

```
irb(main):084:0> [1,2,1,1,3,4,5,6,67,7,7,4,2,342,34,34,432,14].count(1)
=> 3
```

#### 18. it should return an enumerator when no blocks or arguments are given

File: src/enumerable.rb [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  #   result
+  # end
+
+  def my_map
+    result = []
+    my_each { |item| result << yield(item) }
+    result
+  end
+
+  def my_inject(acc = 0)
```
Comment (samgaco) [inline src/enumerable.rb]: #### 19.  should be able to combine all the elements of a collection with a given operation:

```
irb(main):095:0> (1..2).inject { |acc, n| acc + n }
=> 3
irb(main):096:0> (1..2).inject(1) { |acc, n| acc + n }
=> 4
irb(main):097:0> (1..2).inject(10) { |acc, n| acc + n }
=> 13
irb(main):098:0> (1..2).inject(100) { |acc, n| acc + n }
=> 103
```

#### 20. Should be able to perform such operation is passed a symbol as an argument:

```
irb(main):099:0> (1..2).inject(:+)
=> 3
irb(main):103:0> (2..4).inject(:*)
=> 24
```

**Review by samgaco — CHANGES_REQUESTED** [review]
Hi @ferreirati :high_brightness: 

I left 20 points to address in the comments below, please :pray:  work on it and submit back for the release 0.0.1 

You can check some documentation on enumerables [here](https://ruby-doc.org/core-2.5.0/Enumerable.html). Cool readme file by the way. 

Good spirit :100: 

--------------------
* ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

**Comment by tiagoit** [comment]
> 
> 
> Hi @ferreirati 🔆
> 
> I left 20 points to address in the comments below, please 🙏 work on it and submit back for the release 0.0.1
> 
> You can check some documentation on enumerables [here](https://ruby-doc.org/core-2.5.0/Enumerable.html). Cool readme file by the way.
> 
> Good spirit 💯
> 
>     * ###### STATUS OF THE PROJECT: REQUESTED CHANGES 🛑

Hi Sam, I really appreciate this code review, this time I tried to follow the official doc behavior for all the module methods.

File: src/enumerable.rb [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+    selected = []
+    my_each { |obj| selected << obj if yield(obj) }
+    selected
+  end
+
+  def my_all?(pattern = nil)
+    if block_given?
+      my_each { |obj| return false unless yield(obj) }
+    elsif pattern
```
Comment (Torres-ssf) [inline src/enumerable.rb:30]: **[REQUIRED]**
When a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
`[1,2,3,4,5].my_all?(5) => returns false`
`[5,5,5,5,5].my_all?(5) => returns true`

File: src/enumerable.rb [hunk 10]
```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    length.times { |i| yield(self[i]) }
+    self
```
Comment (Torres-ssf) [inline src/enumerable.rb]: **[REQUIRED]**
Try another way to create your `each` method. Right now it does not work with range. Range does not have length.
`(10..15).each {|n| print n, ' ' }`
`# prints: 10 11 12 13 14 15`

File: src/enumerable.rb [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    else
+      my_each { |obj| return true if obj }
+    end
+    false
+  end
+
+  def my_none?(pattern = nil)
+    if block_given?
+      my_each { |obj| return false if yield(obj) }
+    elsif pattern
```
Comment (Torres-ssf) [inline src/enumerable.rb:57]: **[REQUIRED]**
When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern.
`[1,2,3,4].my_none?(5) => returns true`
`[1,2,3,4,5].my_none?(5) => returns false`

**Review by Torres-ssf — CHANGES_REQUESTED** [review]
### Hi there @ferreirati  :hand: 
 Nice job implementing these changes :+1: :clap: 
**But we still need to make some changes before merging this PR.**

Please kindly submit for code review again after making the required changes.
Happy coding :computer:

**Comment by tiagoit** [comment]
> 
> ### Hi there @ferreirati ✋
> 
> Nice job implementing these changes 👍 👏
> **But we still need to make some changes before merging this PR.**
> 
> Please kindly submit for code review again after making the required changes.
> Happy coding 💻

Thanks for the review Torres, all fixed.

File: src/enumerable.rb [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+      my_each do |obj|
+        return false if pattern_match?(obj, pattern)
+      end
+    else
+      my_each { |obj| return false if obj }
+    end
+    true
+  end
```
Comment (maelfosso) [inline src/enumerable.rb:65]:
```ruby
irb(main):015:0> words = %w[dog door rod blade]
=> ["dog", "door", "rod", "blade"]
irb(main):016:0> words[0] = 5
=> 5
irb(main):017:0> words.none?(5)
=> false
irb(main):018:0> words.my_none?(5)
=> true
irb(main):019:0> 
```

File: src/enumerable.rb [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+    array = initial ? self : self[1..-1]
+    initial ||= self[0]
+    if block_given?
+      array.my_each { |item| initial = yield(initial, item) }
+    elsif sym
+      array.my_each { |item| initial = initial.public_send(sym, item) }
+    end
+    initial
+  end
```
Comment (maelfosso) [inline src/enumerable.rb]:
```ruby
irb(main):022:0> range = Range.new(5, 50)
=> 5..50
irb(main):023:0> range.inject(4) { |prod, n| prod * n }
=> 5069015533618896340602101361010794807396273594826752000000000000
irb(main):024:0> range.my_inject(4) { |prod, n| prod * n }
Traceback (most recent call last):
        4: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        3: from (irb):24
        2: from src/enumerable.rb:112:in `my_inject'
        1: from src/enumerable.rb:8:in `my_each'
NameError (undefined local variable or method `length' for 5..50:Range)
irb(main):025:0> 
```

Comment (maelfosso) [reply src/enumerable.rb]:
```ruby
irb(main):028:0> range = Range.new(5, 50)
=> 5..50
irb(main):029:0> range.inject(2, :*)
=> 2534507766809448170301050680505397403698136797413376000000000000
irb(main):030:0> range.my_inject(2, :*)
Traceback (most recent call last):
        4: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        3: from (irb):30
        2: from src/enumerable.rb:114:in `my_inject'
        1: from src/enumerable.rb:8:in `my_each'
NameError (undefined local variable or method `length' for 5..50:Range)
irb(main):031:0> 
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello,

You did a really nice job :smile: 
Thank you for the improvements. 
However, there are still some bug to fix. Please, read the comments below.

Make the changes and submit a new code review request.
Thanks.

**Comment by tiagoit** [comment]
> 
> 
> Hello,
> 
> You did a really nice job 😄
> Thank you for the improvements.
> However, there are still some bug to fix. Please, read the comments below.
> 
> Make the changes and submit a new code review request.
> Thanks.

Hi Fosso, thanks for the review, but I guess you got an outdated commit, can u double check please, all your fixes seems working here.

**Review by Ispirett — APPROVED** [review]
Great job man, you  can merge now :+1:  :100:

**Pull request merged**